testing websocket-core

gregw opened the pull request:
@joakime,

This is my branch for testing websocket-core and trying out some ideas.

My intention is to write the minimal code to make websocket-core be able to accept and/or create websocket connections.

Currently I've been experimenting in `websocket-core/test`, but if we follow this direction the inner classes/interfaces I've created will have to move to either `websocket-core-server/main` / `websocket-core-client/main` or to `websocket-core/main/oejw/core/server`/`websocket-core/main/oejw/core/client`. Let's defer that discussion until the general approach has been agreed.

All in all, it is good news and I've been able to write a little example `org.eclipse.jetty.websocket.testimpl.WebSocketHandler` that is currently able to start a server and accept a ws connection created from the chrome console (but I don't have the chrome-foo yet to work out if it actually handles frames).

To do so, I've had to move down and morph several concepts that exist higher up in the API layers:
- Handshaker
- ConnectionFactory
- SessionFactory

These should be mostly independent of the API layers... with perhaps the exception of the SessionFactory (which IS the API layer as far as the core is concerned).  Note that while I moved these abstractions down, most of the core implementation of them (eg extension negotiation) already existed in core, so it was really just a matter of scrapping off the unnecessary API bits).

Note the pattern where I'm allowing the factories to be set on the context and/or connector, which makes it a little bit more like normal jetty a ConnectionFactory, which are set on connectors and use the connectors resources (buffer pools, executors).    But session factories will most likely be on contexts.... perhaps both will be depending on how/where the policy is configured... and CDI stuff as well.

The biggest area of work need is still around the session.   It still looks way too complex for a core session and I think there a lots of things like upgrade request/response and endpoint objects that can be dropped from the core abstraction.

Anyway, have a look, read the TODOs and lets hangout soon to discuss.  

@sbordet  your input/review also welcome

gregw commented:
@joakime I've updated and now I can send and receive text frames... however I've not got the demasking working correctly for received frames?

joakime commented on jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/WebSocketLocalEndpoint.java:
```diff
@@ -28,25 +28,24 @@
 public interface WebSocketLocalEndpoint
 {
-    Logger getLog();
+    Logger getLog(); // TODO why?
```
This make the logs appear as named from the users endpoint.

gregw replied:
Need convincing, but minor point.

joakime commented on jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/WebSocketLocalEndpoint.java:
```diff
@@ -28,25 +28,24 @@
 {
-    Logger getLog();
+    Logger getLog(); // TODO why?
+    boolean isOpen(); // TODO Does the endpoint really know this?
```
Yes, JSR356 has this requirement.

joakime replied:
I will endeavor to remove this at the websocket-core level.

gregw replied:
exactly, JSR356 stuff should be in the JSR356 API layer

joakime commented on jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/io/WebSocketCoreConnection.java:
```diff
@@ -122,8 +122,8 @@ public WebSocketCoreConnection(EndPoint endp, Executor executor, ByteBufferPool
-        Objects.requireNonNull(upgradeRequest, "UpgradeRequest");
-        Objects.requireNonNull(upgradeResponse, "UpgradeResponse");
+        // TODO Objects.requireNonNull(upgradeRequest, "UpgradeRequest");
+        // TODO Objects.requireNonNull(upgradeResponse, "UpgradeResponse");
```
This is important, upgrade request and response are required.

gregw replied:
They are required by some API layers, but I can't see how they are needed by core.  In fact core is working fine with them null!
I understand that the abstraction of the requests needs to be carried by the session, but that can be added in the API layer as either a session wrapper or extension in a way that is specific to the API layer.

joakime commented on jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/io/WebSocketCoreConnection.java:
```diff
@@ -163,6 +163,11 @@ public void setSession(WebSocketCoreSession session)
+    {
+        return extensionStack;
+    }
+
```
Access to extension stack is NOT required and discouraged

gregw replied:
I needed that at one stage when I was passing just the connection to something that completed the response handshake.  I agree that access to the stack should not be necessary, but getting the negotiated extensions is a reasonable getter to have.  However I think I currently don't need either.

joakime commented on jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/util/QuoteUtil.java:
```diff
@@ -34,6 +34,7 @@
  * <li>Future HTTPbis ABNF changes for parsing will impact QuotedStringTokenizer</li>
  * </ul>
  * It was decided to keep this implementation separate for the above reasons.
+ * @Deprecated use QuotedCSV
```
QuotedCSV cannot handle the quirks in the RFC6455 spec.

gregw replied:
Really?   My preference would be to update or extend QuotedCSV rather than have an entirely different duplicated parser... unless the requirements are significantly different.

gregw replied:
@joakime  QuotedCSV is a much newer class that the Quoted tokenizers that existed when you probably wrote QuoteUtil.   I've had a look at RFC6455 and I think it is substantially in line with RFC7230, which is what QuotedCSV is written against.   Specifically both are written against the #() and *() BNF rules and thus have similar treatment of LWS etc.

I'm confident that QuotedCSV is substantially the same and if any quirks do emerge (which I can't initially see), then it can be modified to cope.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class WebSocketHandler extends HandlerWrapper
```
Do not add this to websocket-core.

This is better served in websocket-servlet

joakime replied:
There is also already a WebSocketHandler in the jetty-websocket-server module.
It handles the Jetty Native WebSocket API features in Handler form.

joakime replied:
Actual testing of the IO in websocket-core is undesired at this moment.

gregw replied:
I didn't use the WebSocketHandler from `jetty-websocket-server` precisely because it handles the jetty native API.  The idea is to test the core layer with zero API dependencies so we can get down to the absolute core mechanism.

Perhaps `websocket-servlet` would be better named `websocket-core-server`.  However, I remain unconvinced that we need to divide the mechanism into many fine grade modules.    A `websocket-core` with a few client and server classes and provided dependencies may be a simpler approach.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+    }
+
+
+    private final static class RFC6455Handshaker implements Handshaker
```
This is what websocket-servlet does

gregw replied:
OK - missed that, but a whole package for one class could be bit heavy.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+
+        final static int VERSION = 13;
+
+        public boolean upgradeRequest(Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException
```
This is what WebSocketServerFactory does (in websocket-servlet)

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+            boolean upgrade = false;
+            QuotedCSV connectionCSVs = null;
+            String key = null;
+            QuotedCSV extensionCSVs = null;
```
This class is does not work for the extension parsing.  The autobahn tests will fail when using QuotedCSV.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+                {
+                    switch(field.getHeader())
+                    {
+                        case UPGRADE:
```
This is a non-optional validation, doing this in a switch is aweful.

gregw replied:
I don't understand this comment.   This loop/switch is done for efficiency (as fields is a list not a map internally) and should do full validation of the required headers... although I currently do a separate getHeader for the version so the code can be made more modular, so we waste one iteration through the headers.... but better than doing 5 or 6.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+                    switch(field.getHeader())
+                    {
+                        case UPGRADE:
+                            if (!"websocket".equalsIgnoreCase(field.getValue()))
```
Upgrade header can contain 2 entries, not just 1.

gregw replied:
ah OK, another CSV then.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+                        case SEC_WEBSOCKET_EXTENSIONS:
+                            if (extensionCSVs==null)
+                                extensionCSVs = new QuotedCSV();
+                            extensionCSVs.addValue(field.getValue());
```
Broken

gregw replied:
As above, I'd prefer to update/extend the utility class rather than duplicate.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+                            break;
+
+                        case SEC_WEBSOCKET_VERSION:
+                            if (field.getIntValue()!=VERSION)
```
This is also comma delimited

gregw replied:
The response version is a list?  wow will have to look up that semantic.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+
+
+
+    static class TestWebSocketConnectionFactory extends ContainerLifeCycle implements WebSocketConnectionFactory
```
WebSocketConnectionFactory ??

That is not required.
Each implementation will have its own raw creation of a Connection.

gregw replied:
Why?   I can see nothing in the connection that is implementation specific.   The connection can be concrete and in core.   Only the session needs to be application layer specific

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+    }
+
+
+    static class TestWebSocketSessionFactory extends ContainerLifeCycle implements WebSocketSessionFactory
```
WebSocketSessionFactory is not used and has been removed recently.

joakime commented on jetty-websocket/websocket-core/pom.xml:
```diff
@@ -26,6 +26,12 @@
     </dependency>
     <dependency>
+      <groupId>org.eclipse.jetty</groupId>
+      <artifactId>jetty-server</artifactId>
```
No.  This is what websocket-servlet is for.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+    private final static class RFC6455Handshaker implements Handshaker
+    {
+
+        final static int VERSION = 13;
```
Use WebSocketConstants.SPEC_VERSION

gregw replied:
OK

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+                    LOG.debug("not upgraded no session {}",baseRequest);
+            }
+
+            connection.setSession(session);
```
This is incorrect.

The `session.setWebSocketEndpoint(Object endpoint, WebSocketPolicy policy, L localEndpoint, R remoteEndpoint)` needs to be called instead.

joakime replied:
While the connection needs access to the session, the way the other 4 APIs are doing this now is via the creation of the connection (which needs a session), extend WebSocketCoreConnection and have its constructor call .setSession(connection) instead.

joakime replied:
The connection uses the session for ...

* The destination of parsed frames (after the extensionstack) `session.incomingFrames()`
* Reporting the Jetty `Connection.onOpen()` to the RFC6455 OPEN phase `session.open()`, which triggers the `localEndpoint.onOpen()`
* Reporting Connection level errors via the `session.processError(Throwable)`

To eliminate this ...

* Using `Connection.Listener.onOpened()` can be used for the `session.open()`
* Passing `IncomingFrames` instead of Session to the connection perhaps like `connection.setIncoming(session)`
* Using an new interface such as `ConnectionErrorListener.onConnectionError(Throwable t)` which the session implements instead.

gregw replied:
The `session.setWebSocketEndpoint(Object endpoint, WebSocketPolicy policy, L localEndpoint, R remoteEndpoint)` is done in my test code within the session factory.   The core code doesn't need to know how to construct a session, it only needs to link the abstract session to the connection, which is exactly what `connection.setSession(session)` does.

joakime commented on jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/WebSocketLocalEndpoint.java:
```diff
@@ -28,25 +28,24 @@
+    default void onError(Throwable cause) {}
 
-    void onText(Frame frame, Callback callback);
+    default void onText(Frame frame, Callback callback) {}
```
I disagree with defaults, this is not an API for users.
This is a core level API for other APIs to build off of, each of those APIs needs to implement each of these interfaces so they understand what is required of them.

gregw replied:
OK.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+
+    public interface Handshaker
+    {
+        boolean upgradeRequest(Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException;
```
Use UpgradeRequest and UpgradeResponse, not Request, HttpServletRequest, and HttpServletResponse.

The UpgradeRequest and UpgradeResponse objects are suitable for being used by APIs and users as those do not get recycled by Jetty.  
The Jetty Request object is not exposed via the WebAppClassLoader.

gregw replied:
I don't see why core needs UpgreadeRequest and UpgradeResponse.  Core can definitely work at with the jetty APIs and we are always be able to discover the core Request object even from a wrapped ServletRequest.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+        boolean upgradeRequest(Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException;
+    }
+
+    interface WebSocketConnectionFactory extends ConnectionFactory
```
A ConnectionFactory for WebSocket's would need to interrogate the user provided endpoint to function (as some features of the connection, such as the subprotocol negotiation, and extension negotiation, both require knowledge from an endpoint instance)

This concept from the Connectors in Jetty will be difficult to implement.

gregw replied:
Yep - just trying it out for size to get a feel for what needs to be where.   For example, if the subprotocol is from the application, then it should be negotiated in the session and all that core needs is the ability to pass the candidate subprotocols to the session factory and then get the resulting subprotocol from the session to use in the handshake response. 

If that is the case, then subprotocols should not be configured on the connection factory and it still may be possible to have a relationship with a connector....  however, I'm prepared to concede this might be a step too far.   But let's see if it is possible.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+
+    interface WebSocketSessionFactory
+    {
+        // TODO why is core session abstract?
```
WebSocketCoreSession is not exposed via the WebAppClassLoader, it is not an end user API, the specific APIs need to extend from WebSocketCoreSession to make their own APIs work successfully through the WebAppClassLoader.

gregw replied:
They may extend, they may wrap, they may map.  The APIs should be free to use the core session however they want.    There are no abstract methods on core and it appears functional as is, so it is not abstract.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+    interface WebSocketSessionFactory
+    {
+        // TODO why is core session abstract?
+        WebSocketCoreSession<ContainerLifeCycle,WebSocketCoreConnection,WebSocketLocalEndpoint,WebSocketRemoteEndpoint>
```
You should extend WebSocketLocalEndpoint and WebSocketRemoteEndpoint for your new API.

gregw replied:
Yes for `WebSocketLocalEndpoint` which is abstracted and given to the core session. but I think no so much for `WebSocketRemoteEndpoint` which should be provided by the core session.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+
+    protected Handshaker getHandshaker(Request baseRequest)
+    {
+        // TODO this can eventually be made pluggable
```
Already done in WebSocketServerFactory and WebSocketServletFactory

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+            baseResponse.getHttpFields().add(ConnectionUpgrade);
+            baseResponse.getHttpFields().add(HttpHeader.SEC_WEBSOCKET_ACCEPT, AcceptHash.hashKey(key));
+            baseResponse.getHttpFields().add(HttpHeader.SEC_WEBSOCKET_EXTENSIONS,
+                                             ExtensionConfig.toHeaderValue(connection.getExtensionStack().getNegotiatedExtensions()));
```
If you need the connection, then you are doing this in the wrong order.
The upgrade occurs before the websocket specific connection is even created.

sbordet replied:
@joakime the upgrade cannot occur before the WebSocket connection is created, as we should call `endPoint.upgrade(Connection)` and pass it the new connection.

joakime replied:
ExtensionStack is the post-negotiated reality of the active connection.  
Using it to perform the negotiation is incorrect.

gregw replied:
I don't think the implementation matters either way.  We can send upgrade response and then do the upgrade or the other way around, because we are not fill interested during this period.

I did it this way just so that if there is any error in generating the upgrade response then we do not do the upgrade.

See comments above regarding extension stack

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+
+            // TODO why is remoteEndpoint an interface rather than just an impl?
+            WebSocketRemoteEndpointImpl remoteEndpoint = new WebSocketRemoteEndpointImpl(connection);
+            remoteEndpoint.open(); // TODO why is this needed?
```
This not done by you and your session implementation.
The .open() call is for when the handshake is complete and the RFC6455 state changes to OPEN.

gregw replied:
Hmm wasn't happening, so need to debug that.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+        {
+
+            // TODO why is remoteEndpoint an interface rather than just an impl?
+            WebSocketRemoteEndpointImpl remoteEndpoint = new WebSocketRemoteEndpointImpl(connection);
```
Interface is easier to test with, the default implementation is present in the *Impl() class.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+            remoteEndpoint.open(); // TODO why is this needed?
+
+            // TODO abstract the creation of a local Endpoint
+            WebSocketLocalEndpoint localEndpoint = new WebSocketLocalEndpoint()
```
You don't create a WebSocketLocalEndpoint, that's not a User API.

In a server usage, use the WebSocketCreator to obtain an API specific endpoint instance, then a factory to create/manage the metadata for that instance, and the factory is then capable of providing WebSocketLocalEndpoint instances from the metadata that is bound to that specific instance.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+                    text.setPayload("Opened!");
+
+                    // TODO do I need to mask this myself? if so then why?
+                    remoteEndpoint.sendFrame(text, new Callback()
```
You should be using Session.getRemote().sendText().

The WebSocketRemoteEndpoint is not an end user API.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+                }
+
+                @Override
+                public void onOpen()
```
Where's your Session?

gregw replied:
In this case, my test "API" layer has no requirement on the session beyond what is provided by the core, so the core session is just fine and does not need to be extended/wrapped/mapped etc.  In fact the core session already does too much.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+                @Override
+                public boolean isOpen()
+                {
+                    return connection.isOpen();
```
Connection is insufficient to tell if "isOpen" is valid.
The flusher state and RFC6455 state must also be interrogated.

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+                @Override
+                public void onText(Frame frame, Callback callback)
+                {
+                    ByteBuffer payload = frame.getPayload();
```
Again, WebSocketLocalEndpoint is not an end user API.
You shouldn't be using it this way.

gregw replied:
I'm pretending to be then end user API.  So I'm doing what an end user API might do to get to the payload.     If I shouldn't be doing something, then the core API should not let me do it!

joakime commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+                    };
+
+            //TODO Why can't this be done in constructor?
+            //TODO Why does session need the endpoint object?
```
Order of creation is different in the different APIs and modes (server vs client).

gregw replied:
Ah OK - need to test client side as well then to get a feel for that.  yuck I hate clients!

gregw commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java:
```diff
@@ -0,0 +1,457 @@
+                        case SEC_WEBSOCKET_SUBPROTOCOL:
+                            if (subprotocol!=null)
+                                return false;
+                            subprotocol = field.getValue();
```
OK, so we need to feed the list to the session negotiation and then ask the resulting session for the single one that was negotiated?  or is the response also a list?

gregw commented:
@joakime @sbordet   I'm in a moderately happy place with the direction of this branch at the moment.  However, as previously discussed the creation and wiring up of the connection and session is still the main issue to be resolved.

Currently this test does:
1. Negotiate the extensions
2. Create the Connection
3. Create the Session(passing the Connection)
4. Wire the session to the connection with a setter

This works except for the key problem that some APIs allow the endpoint to be involved with the negotiation of extensions.   Currently we do that in the API by splitting the creation of the endpoint object from the creation of the session and putting the extension negotiation and connection creation in between.  However, from core's point of view, there is no endpoint, only a session (and some APIs might not have an endpoint concept), so there is no way to do that same split.

We could try
1. Create the Session
2. Negotiate the extensions
3. Create the Connection (passing in the session)
4. Wire the connection to the session with a setter

Which will need a little bit of radical surgery, but I think it is probably a more logic order, as we should not create the connection unless we know the API layer is prepared to accept that particular endpoint.

So far I have refrained from radical surgery in core, but I may give that a go over the next few days just to see what it looks like.... but still consider it experimental rather than intended to be merged.

I will also get a bit more aggressive about stripping creating a minimal session.

gregw commented:
My latest commits have been brutal cut backs of the code in core, to just what I think is necessary.  I didn't move code out to other modules... just deleted it, even though I know some of it will be necessary in higher level modules.

I think from a simplicity point of view, websocket-core is pretty close to where it should be.  I've reworked the lifecycle so the session is created first and it can influence the policy, the extensions and the protocol.  However, I've not yet done the client side, so a few adjustments are needed there.

@sbordet I'm a bit concerned about the whole `isOpen`, closing and error shutdown code, so it would be great if you could review that and we can all get together early next week to discuss.

@joakime I think you can start looking at this from the point of view of implementing the jetty API and/or the RFC API and telling me what I've cut out of core that really is needed.     

One area I think I've probably got wrong is the use of the object decorator.  Currently core only decorates extensions that are created, plus it delegates the creation of the session and the localendpoint, so both of those can be decorated by an impl.       However, I think we need a described policy of exactly what objects should and should not be decoratable.

I'll have a go at a simple core client early next week.

gregw commented:
@sbordet @joakime following Simone's review I have moved the ExtensionStack to be part of the session rather than the connection.   However, I couldn't bring myself to also move the flusher and generator there.

The connection now has Parser and its handler inbound and flusher and generator outbound, so it feels nicely symmetric.

The session has the extension stack which is wired up to the local and remote endpoints at the top of the stack and to the connection at the bottom of the stack.

I've also removed several close closing closeSent AtomicBooleans, as such multiple booleans are never actually atomic.  Instead I'm currently just using the ConnectionState for all the decisions.... however there are close problems which I need to talk through with you both before so we can decide what to do.

First problem is that the previous code appeared to switch to CLOSED state both when a CloseFrame had been successfully sent and when a CloseFrame was received in reply to a CloseFrame sent.  I assume we want one behaviour or the other, not both?

I've tried to do the first, but that means that from the close frame sent succeeded callback I am calling disconnect to close the flusher, which is still processing as it is sending the close frame... which causes a verbose exception.  I think that this exception was happening before, but was just ignored.

Finally, in the session, I can't tell if a Close frame received is in response to a Close frame sent or not. That state is in the connection.  So perhaps all of the close state should be in the session, which probably means we really don't need ConnectionState anymore as a class/concept?

joakime commented:
![websocket core / refactor](https://user-images.githubusercontent.com/191721/31471715-1576a01a-ae9f-11e7-8485-504133730b9d.png)

This is how the layers fit together.
I disagree with the move of ExtensionStack to session. 
The ExtensionStack is part of the HTTP upgrade/negotiate phase/scope.
The Connection is charged with holding all of the immutable state/config from the negotiate (the sub-protocols, realized extensionstack, request state, response state) along with doing the bytebuffer <-> frame conversions (parser and generator)
The Session only deals with Frames.
FrameFlusher exists after the RemoteEndpoint and before the ExtensionStack.

gregw commented:
Joakim,

I was sceptical too about moving it, but it did simplify a lot of things:

   - It is the session that is part of the extension negotiation
   - The code that was inserted between the stack and the flusher was
   referring to the session anyway, so that code is now just in the session.

However, I'm not sure the division is exactly settled yet.  In fact I now
think that the ConnectionState should become a SessionState and that the
connection itself will become stateless and just a meeting point of
endpoints, parsers, generators and sessions - all of which look after their
own state.

But yes - let's find a time for the 3 of us to review this and get an
agreement between Simone (put it all in the session), you (put it back in
the connection) and me (doesn't Goldilocks have a solution to this?)

cheers



On 12 October 2017 at 10:18, Joakim Erdfelt <notifications@github.com>
wrote:

> [image: websocket core / refactor]
> <https://user-images.githubusercontent.com/191721/31471715-1576a01a-ae9f-11e7-8485-504133730b9d.png>
>
> This is how the layers fit together.
> I disagree with the move of ExtensionStack to session.
> The ExtensionStack is part of the HTTP upgrade/negotiate phase/scope.
> The Connection is charged with holding all of the immutable state/config
> from the negotiate (the sub-protocols, realized extensionstack, request
> state, response state) along with doing the bytebuffer <-> frame
> conversions (parser and generator)
> The Session only deals with Frames.
> FrameFlusher exists after the RemoteEndpoint and before the ExtensionStack.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/jetty.project/pull/1866#issuecomment-335975608>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEUrVnO-Cl4rC573rdm6uthxSaWVtJEks5srUzZgaJpZM4PtB4B>
> .
>



-- 
Greg Wilkins <gregw@webtide.com> CTO http://webtide.com

gregw commented:
I'm AFK for a few hours, but will draw up a pretty diagram like yours for
how I see the split.

cheers


On 12 October 2017 at 10:26, Greg Wilkins <gregw@webtide.com> wrote:

> Joakim,
>
> I was sceptical too about moving it, but it did simplify a lot of things:
>
>    - It is the session that is part of the extension negotiation
>    - The code that was inserted between the stack and the flusher was
>    referring to the session anyway, so that code is now just in the session.
>
> However, I'm not sure the division is exactly settled yet.  In fact I now
> think that the ConnectionState should become a SessionState and that the
> connection itself will become stateless and just a meeting point of
> endpoints, parsers, generators and sessions - all of which look after their
> own state.
>
> But yes - let's find a time for the 3 of us to review this and get an
> agreement between Simone (put it all in the session), you (put it back in
> the connection) and me (doesn't Goldilocks have a solution to this?)
>
> cheers
>
>
>
> On 12 October 2017 at 10:18, Joakim Erdfelt <notifications@github.com>
> wrote:
>
>> [image: websocket core / refactor]
>> <https://user-images.githubusercontent.com/191721/31471715-1576a01a-ae9f-11e7-8485-504133730b9d.png>
>>
>> This is how the layers fit together.
>> I disagree with the move of ExtensionStack to session.
>> The ExtensionStack is part of the HTTP upgrade/negotiate phase/scope.
>> The Connection is charged with holding all of the immutable state/config
>> from the negotiate (the sub-protocols, realized extensionstack, request
>> state, response state) along with doing the bytebuffer <-> frame
>> conversions (parser and generator)
>> The Session only deals with Frames.
>> FrameFlusher exists after the RemoteEndpoint and before the
>> ExtensionStack.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/eclipse/jetty.project/pull/1866#issuecomment-335975608>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAEUrVnO-Cl4rC573rdm6uthxSaWVtJEks5srUzZgaJpZM4PtB4B>
>> .
>>
>
>
>
> --
> Greg Wilkins <gregw@webtide.com> CTO http://webtide.com
>



-- 
Greg Wilkins <gregw@webtide.com> CTO http://webtide.com

gregw commented:
So below is my current vision of what the break up should be.  Note that currently the branch is not like this as the SessionState is still ConnectionState within the connection, so you have the far side of the extension in stack reaching into the connection to twiddle state.   More notes below:

![gregs vision of websocket core 2f refactor](https://user-images.githubusercontent.com/70829/31478629-6987f6f4-af5d-11e7-99eb-188f5ad82f5e.png)

So the key part about this break down is that it needs and IncomingState and OutGoingState object to cap either side of the extension stack so that Close frames can be detected/handled/injected/responded to.   I have Close frames going via the extension stack, so it is possible for extensions to intercept Close frame and to do funky stuff like injecting more frames before finally passing on the CloseFrame (it can't eat it forever).

The OutGoingState object is effectively a wrapper of the Flusher that converts the OutgoingFrame API to the Flusher API, so it could have been the Flusher and the Flusher could have been in the session... but it would still have had to be extended.  By having the Flusher in the connection and the OutGoingState wrapper in the session, we achieve a nice symmetry that the connection just produces and consumes Frames without state.  I'm thinking of Google here, in that if we replaced the Connection with an RPC layer, we would want to replace the Flusher - so it should not be in the Session!

I've shown Session extending IncomingFrames and Connection extending OutgoingFrames... but I'm not totally convinced on that yet.  It does save wiring the Connection>Parser>Handler directly to the Session>ExtensionStack   and Session>OutgoingState directly to the Connection>Flusher, but perhaps it buys us more confusion as it gives more ways to plug this all together wrongly?

I'll now have a go at making the branch match that diagram.

gregw commented:
Branch has been updated to match my diagram.
It works pretty well and is a close match to the HTTP stack:
 - WebsocketCoreConnection == HttpConnection
 - WebsocketCoreSession == HttpChannel
 - WebsocketSessionState == HttpChannelState

joakime commented:
I've merged your changes.

Initial impressions: 
This makes no sense anymore.
The entire /cdi-websocket/ layer has to be removed, this new reality makes it impossible as it stands (the changes in websocket-core mean that CDI now has to be implemented at each API level: jetty native client, jetty native servlet, jetty native handler, jetty native filter, jsr356 client, jsr356 server, jsr356 client on server.)  This is because there's no clear separation of the scopes anymore at websocket-core: we need HTTP (application scope, servlet scope, request scope, upgrade scope) and WebSocket (container scope, handshake scope, session scope, message scope)
The amount of work pushed up into APIs has quadrupled in complexity and duplication (this screams of more bugs and higher maintenance demands)

gregw commented:
Joakim,

I fail to see how having a simply core layer that can run a websocket
connection free from any API complexity can result in more bugs or
maintenance problems.   As it is the biggest problems with websockets have
been stalls and bad closes (just like http/1 and http/2) and removing the
API complexity is key to getting the core connection right!

I do get it that the API layers need more complications and they need to
have state/info held by and transferred by the core layer.  But let's
explicitly identify what those are and provide minimal abstractions for
them.  We always have the common layer to put more shared stuff if there
are other non core shared abstractions.

cheers




On 13 October 2017 at 05:51, Joakim Erdfelt <notifications@github.com>
wrote:

> I've merged your changes.
>
> Initial impressions:
> This makes no sense anymore.
> The entire /cdi-websocket/ layer has to be removed, this new reality makes
> it impossible as it stands (the changes in websocket-core mean that CDI now
> has to be implemented at each API level: jetty native client, jetty native
> servlet, jetty native handler, jetty native filter, jsr356 client, jsr356
> server, jsr356 client on server.) This is because there's no clear
> separation of the scopes anymore at websocket-core: we need HTTP
> (application scope, servlet scope, request scope, upgrade scope) and
> WebSocket (container scope, handshake scope, session scope, message scope)
> The amount of work pushed up into APIs has quadrupled in complexity and
> duplication (this screams of more bugs and higher maintenance demands)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/jetty.project/pull/1866#issuecomment-336231733>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEUrVX6I6bUa4c5jdXZRIlQvbhwEl0vks5srl_KgaJpZM4PtB4B>
> .
>



-- 
Greg Wilkins <gregw@webtide.com> CTO http://webtide.com

gregw commented:
@sbordet I have to run, so finishing our conversation here.

You want us to change:
```
class ExampleLocalEndPoint
{
   void onOpen(WebsocketCoreSession session);
}
```

to

```
class ExampleLocalEndPoint
{
   void setSession(WebsocketCoreSession session);
   void onOpen();
}
```

because some API layers may have need of the session before they received the onOpen event.

However, that is exactly the kind on API behaviour that the core layer has no need to know about.  Specifically if we did the change like you suggest, then the implementation inside the core session would just be:
```
    // Open the session
   localEndPoint.setSession(this)
   localEndPoint.open();
```

To be precise, the core implementation does not have any semantic it needs/wants/supports on the local endpoint that has the session set on it, but open not yet called.  So if we go back to
```
   localEndPoint.open(session);
```
Then it is entirely appropriate for an API adaption layer to implement the LocalEndPoint (which is the core endpoint and NOT the API endpoint) as something like:
```
   class MyApiWebSocketLocalEndPoint
   {
        WebsocketCoreSession session;
        MyApiEndPoint endpoint;
  
        onOpen(WebSocketCoreSession session)
        {
             this.session = session;
             endp.setSession(session);
             // do something 
             endp.onOpen();
        }
}
```

So the semantic and sequence  you wish to achieve can be done without any involvement of websocket-core.

In short core doesn't care about the difference between setSession(session) and onOpen(), so there does not need to be the two methods.   You are confusing the ultimate API endpoint with the endpoint that the core needs.... perhaps it should not be call END point

sbordet commented:
@gregw there definitely exist an "session open" event that needs to be delivered to applications, in general, no matter the API you use.

LocalEndPoint is the recipient of that event, and has the responsibility of converting that event to something specific for a given API (e.g. calling an annotated method). 

To me the event has nothing to do with wiring objects together, like your `onOpen()` does.
Wiring objects must happen _before_ the "session open" event is delivered.

Sure you can squash wiring objects and event notification in one method, but IMHO it's important to keep the two separated because the wiring is implementation specific (i.e. does not appear in the `LocalEndPoint` interface), while the event notification is generic (i.e. it _does_ appear in the `LocalEndPoint` interface):

```java
interface LocalEndPoint {
    void onOpen();
    ...
}

class JSRLocalEndPoint implements LocalEndPoint {
    void setSession(JSRSession s) {
        this.session = s;
    }
    
    @Override
    void onOpen() {
        ...
    }
}
```

joakime commented:
if the following is used ...

``` java
class ExampleLocalEndPoint
{
   void onOpen(WebsocketCoreSession session);
}
```

then the API's would need to extend the WebSocketCoreSession to add the functionality they need, resulting in WebSocketCoreSession needing to be refactored to not take LocalEndPoint or RemoteEndPoint in their constructors, they will need to be passed in.

API would use them this way ...

``` java
class APISession extends WebSocketCoreSession
{
  public APISession(APIContainer container, APIConnection apiConnection,
       APIUpgradeRequest request, APIUpgradeResponse response, 
       Map<String,String> pathMappingParams, WebSocketPolicy sessionPolicy)
  {
  }
}

APISession session = new APISession(...);
Object userProvidedPojo = getUserPojoInstance();
LocalEndPointAPIMetadata localEndpointMetadata = localEndpointFactory.getMetadata(userProvidedPojo.getClass());
LocalEndPointAPIImpl localEndpoint = localEndpointMetadata.realize(userProvidedPojo, session);
RemoteEndpointAPIImpl remoteEndpoint = new RemoteEndpointAPIImpl(session);
session.setEndPoints(localEndpoint, remoteEndpoint);
```

joakime commented:
Regarding RFC6455 and OPEN state...

The states on RFC6455 are important, even to the API level, which uses them and adds a few more important events to track.

The combined states that are important ...

1. You enter CONNECTING when handshake starts. _(RFC specified)_
2. You enter CONNECTED when handshake is complete, but before the local endpoint has been called. _(RFC Specified)_
3. You enter OPENING when local endpoint onOpen() is called. _(API specified)_
4. You enter OPEN when local endpoint onOpen() completed without error (or local endpoint didn't call `.close()`) _(RFC and API specified)_
5. You enter CLOSING when close handshake starts _(RFC specified)_
6. You enter CLOSED when entire websocket is closed from completed close handshake, or due to abnormal close situation. _(RFC specified)_

There are to be no frames read/delivered to the local endpoint until OPEN state has been reached.
Frames can be sent to remote endpoint when OPENING occurs.

In jetty-9.4.x, the `Connection.onOpen()` results in `Session.open()` which results in `LocalEndpoint.onOpen()`

joakime commented:
With discussion with Simone about above states:

The WebSocket states are useful to know and understand, but a rigid/verbatim usage of them in websocket-core isn't likely.  (for example: CLOSING is actually more nuanced about knowledge of local vs remote initiated)
Understanding the points in time for each event/state and what you can/cannot do is the important part, exposing these states in the websocket-core as via methods/API is not useful.

gregw commented:
Simone and Joakim,

You have to look at this from core's point of view.   Is there anything
that a session/endpoint can do between the semantic of injecting the
session and an onOpen call?

Can a message be received? no it is not open!
Can a message be sent? no it is not open!
Can a connection be closed? no it is not open!
Does core do anything else?  no!

So core will never do anything nor allow anything to be done between a
setSession(session) call and an onOpen() call.   API adaption layers are
free to separate those events for their endpoint, but the two events have
no meaning for core and thus should not be separate calls.

The core session and endpoint objects are not the API session and endpoint
objects.

Again I'm thinking that EndPoint may be the wrong name here as they are not
the End.

cheers











On 14 October 2017 at 04:33, Joakim Erdfelt <notifications@github.com>
wrote:

> With discussion with Simone about above states:
>
> The WebSocket states are useful to know and understand, but a
> rigid/verbatim usage of them in websocket-core isn't likely. (for example:
> CLOSING is actually more nuanced about knowledge of local vs remote
> initiated)
> Understanding the points in time for each event/state and what you
> can/cannot do is the important part, exposing these states in the
> websocket-core as via methods/API is not useful.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/jetty.project/pull/1866#issuecomment-336518002>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEUremXudIg958gGNePZwooNXBw72yCks5sr57igaJpZM4PtB4B>
> .
>



-- 
Greg Wilkins <gregw@webtide.com> CTO http://webtide.com

gregw commented:
discussion/refactoring continued in https://github.com/eclipse/jetty.project/pull/1895
